lib/ddtrace/contrib/rails/patcher.rb

```diff
@@ -45,17 +45,26 @@ def before_intialize(app)
         end
 
         def add_middleware(app)
-          # Add trace middleware
+          # Add trace middleware at the top of the middleware stack,
+          # to ensure we capture the complete execution time.
           app.middleware.insert_before(0, Datadog::Contrib::Rack::TraceMiddleware)
 
-          # Insert right after Rails exception handling middleware, because if it's before,
-          # it catches and swallows the error. If it's too far after, custom middleware can find itself
-          # between, and raise exceptions that don't end up getting tagged on the request properly.
-          # e.g lost stack trace.
-          app.middleware.insert_after(
-            ActionDispatch::ShowExceptions,
-            Datadog::Contrib::Rails::ExceptionMiddleware
-          )
+          # Some Rails middleware can swallow an application error, preventing
+          # the error propagation to the encompassing Rack span.
+          #
+          # We insert our own middleware right before these Rails middleware
+          # have a chance to swallow the error.
+          #
+          # Note: because the middleware stack is push/pop, "before" and "after" are reversed
+          # for our use case: we insert ourselves with "after" a middleware to ensure we are
+          # able to pop the request "before" it.
+          if defined?(::ActionDispatch::DebugExceptions)
+            # Rails >= 3.2
+            app.middleware.insert_after(::ActionDispatch::DebugExceptions, Datadog::Contrib::Rails::ExceptionMiddleware)
+          else
+            # Rails < 3.2
```

Comment:
Do we still support Rails < 3.2? Thought we dropped support for this.

Reply:
We still support it for Ruby 2.1: https://github.com/DataDog/dd-trace-rb/blob/a3236ddf0cead1c2a6db57acd9e4ebefbfd45026/Rakefile#L279

But I can see it as a good candidate for deprecation. It would simplify our code base for sure.

Reply:
👍  I think dropping support for Rails 3.1 even before we drop support for Ruby 2.1 makes sense, given that the [3.2 series](https://rubygems.org/gems/rails/versions/3.2.0) still works for 2.1 thus I suspect the intersection of the following groups:

* customers still on 2.1
* customers till on Rails 3.1
* customers that upgrade dd-trace-rb regularly

is probably nil or very close to it.